src/glang/codegen/builtin_types.py

```diff
@@ -472,8 +473,15 @@ def copy_with_storage_kind(self, parent: Type, kind: Type.Kind) -> Type:
         if kind == Type.Kind.VALUE:
             return self.value_type
 
-        assert kind == self.storage_kind
-        return parent
+        if kind == self.storage_kind:
+            return parent
+
+        assert self.storage_kind == Type.Kind.MUTABLE_OR_CONST_REF
+
+        # TODO there must be a better way.
+        new_definition = copy(self)
+        new_definition._storage_kind = kind
+        return AnonymousType(new_definition)
```

Comment:
This is a little scary... This method made our type resolution kinda buggy by copying classes we were referencing, I fixed it by removing all the copies so we could use the `UnresolvedReference` class.... If this causes issues I suppose we're gonna have to rethink how either of these things work.

Reply:
> by copying classes we were referencing

Hmm, not sure I understand why that's problematic. I've changed it to return `AnonymousType(ReferenceDefinition(...))` to avoid the copy. If you still don't like it then open an issue pls